### ISSUE TITLE
Importing relationships via CSV

### ISSUE THREAD

**Issue by FelixMalfait** [twentyhq/core-team-issues#103]
The biggest limitation of our CSV feature today is that it's not possible to use it to create relationships.
So for example if you import a list of people, it can't automatically create/link companies attached to it.

There are different ways to approach this problem.
I think the best way for now would be to allow people to include unique identifiers that would automatically be surfaced.
A unique identified can be: the id, and external if (e.g. SalesforceId), but also sometimes a business key (email for persons or domain for company).

Unfortunately we don't really have a good way to add unicity constraint to a field, and therefore fetch fields that could be good candidates in a good way. So for v1 I suggest we retried id, externalId only for all objects. And hardcode domain/email for company/people.

So on the frontend, for a given object, we need to add to the list of matchable fields all the fields the one that are coming from relations ([id, externalId] in most cases, and that +domain/email for company/person).
And on the backend side, we need to adapt our resolvers to also be able to match through this (company.domain, siblingObject.externalId, etc.)

**Comment by livingstonlarus** [comment]
Hi Felix,

+100 for this if this is what would solve my issue. Coming from Pipedrive, I have 2 CSV files for People and Companies. Companies have an ID column, and People have an Organization ID column.

Maybe it's the `externalId` you are discussing here? That could be the matching key between People and Companies when importing?

**Comment by FelixMalfait** [comment]
@livingstonlarus thanks, we'll make sure to prioritize this in June! yes externalId would be your pipedriveId and that would work

**Comment by raphaelm-gioa** [comment]
Hey! Any news ? Do you have any hints to where this is setup ? I might open a PR if so !

**Comment by FelixMalfait** [comment]
Sorry no I don't ; it's probably not a quick/easy task, it should involve both frontend and backend

**Comment by FelixMalfait** [comment]
Pre-requisite/related: https://github.com/twentyhq/core-team-issues/issues/115

**Comment by celestial-rose** [comment]
Hey there ! I see the sub-issue is completed, was wondering what could be needed to improve this - this is my last step towards migrating to Twenty full on !

**Comment by etiennejouan** [comment]
Hi @celestial-rose! You can now connect (create relationships between objects) through CSV import based on any custom unique field (such as external ID).

To do this, you need to:
- Create a custom unique field on object A that you want to connect to (for example: a new 'externalId' unique text field on Company)
- Import your records for object A with the custom unique field populated (all your companies with externalId)
- Import your records for object B, connecting them to object A based on object A's custom unique field (import a CSV with all your people + the company externalId column)

Please note that you can only connect on the many-to-one side (one-to-many connections are planned for later). 
We will also work on a 'connect or create' feature that will enable you to create the record on the other side of the relationship if it doesn't exist.

**Comment by FelixMalfait** [comment]
Good point this should be closed!

**Comment by celestial-rose** [comment]
Amazing news ! thank you for the thorough explanation, will let you know if I encounter any issues, kudos to your team !

**Comment by FelixMalfait** [comment]
Thank you!!!

**Comment by livingstonlarus** [comment]
Hi! So this resolve the issue I discussed here in May 2024? Now we can work more easily with Pipedrive exports where you have 2 CSV files for People and Companies? Companies have an ID column, and People have an Organization ID column. This custom unique field can be the matching key between People and Companies when importing, where People will have the Company ID field?

If yes, that's a major feature to drive migration to Twenty!

**Comment by FelixMalfait** [comment]
Yes exactly that's it! Getting to a great software step by step - more to come!
